## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#af25308a-129b-497d-946f-339be42b5fc7) |
| mint-compression-dist-xl.sh | :x: [more...](#dcb23c1f-86ee-461f-b624-ded29675795d) |
| mint-compression-fs.sh | :x: [more...](#b90eb17e-dfa8-4080-95a7-a95dd673f54c) |
<a name="af25308a-129b-497d-946f-339be42b5fc7"></a>
### 8510-54e661e/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30846
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c568445c2d75:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 5 seconds
(3/14) Running aws-sdk-php tests ... done in 50 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 2 minutes and 34 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 17 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 42 seconds
(9/14) Running minio-go tests ... FAILED in 5 minutes and 35 seconds
{
  "args": {},
  "duration": 114328,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="b90eb17e-dfa8-4080-95a7-a95dd673f54c"></a>
### 8510-54e661e/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:32431
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 958251c66165:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 35 seconds
(8/14) Running minio-dotnet tests ... done in 32 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 26 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 8 seconds
(11/14) Running minio-js tests ... done in 45 seconds
(12/14) Running minio-py tests ... FAILED in 4 minutes and 0 seconds
{
  "message": "AccessDenied: message: Access Denied",
  "duration": 46,
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1345, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2040, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1359, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "status": "FAIL",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "args": {
    "object_name": "b415474f-acb1-408c-b7c0-20bb1651ff11",
    "presigned_get_object_url": "http://72.28.97.53:32431/minio-py-test-942234cd-8474-4a85-8c15-73c6bb2ef146/b415474f-acb1-408c-b7c0-20bb1651ff11?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191112%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191112T095841Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=7c73b38d76ee47d6d7308cef12cb5625f1854301ad23701aa307b53113c6b2ab",
    "response.status": 403,
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>b415474f-acb1-408c-b7c0-20bb1651ff11</Key><BucketName>minio-py-test-942234cd-8474-4a85-8c15-73c6bb2ef146</BucketName><Resource>/minio-py-test-942234cd-8474-4a85-8c15-73c6bb2ef146/b415474f-acb1-408c-b7c0-20bb1651ff11</Resource><RequestId>15D661E08F0A2443</RequestId><HostId>dfac3a99-58fc-4f10-8566-acc0ffe362b8</HostId></Error>",
    "bucket_name": "minio-py-test-942234cd-8474-4a85-8c15-73c6bb2ef146",
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-11-12T09-29-21Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15D661E08F0A2443\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Tue, 12 Nov 2019 09:58:56 GMT\"]}}",
    "response.reason": "Forbidden"
  }
}

Executed 11 out of 14 tests successfully.
```
<a name="dcb23c1f-86ee-461f-b624-ded29675795d"></a>
### 8510-54e661e/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31726
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1c08546d5518:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 3 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 47 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 2 minutes and 3 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 0 seconds
(8/14) Running minio-dotnet tests ... done in 47 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 23 seconds
(10/14) Running minio-java tests ... done in 3 minutes and 13 seconds
(11/14) Running minio-js tests ... done in 1 minutes and 10 seconds
(12/14) Running minio-py tests ... FAILED in 4 minutes and 46 seconds
{
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "duration": 341,
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1345, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2040, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1359, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "message": "AccessDenied: message: Access Denied",
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "status": "FAIL",
  "args": {
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-11-12T09-29-21Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15D662281E5F85DB\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Tue, 12 Nov 2019 10:04:04 GMT\"]}}",
    "object_name": "2471cd25-f3b6-4845-89ba-31cb16ac3b1d",
    "response.reason": "Forbidden",
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>2471cd25-f3b6-4845-89ba-31cb16ac3b1d</Key><BucketName>minio-py-test-3983798c-873d-45bf-ba41-050b3d8c00d3</BucketName><Resource>/minio-py-test-3983798c-873d-45bf-ba41-050b3d8c00d3/2471cd25-f3b6-4845-89ba-31cb16ac3b1d</Resource><RequestId>15D662281E5F85DB</RequestId><HostId>6d9ac2b6-76fa-4d78-a98f-aea134f7627e</HostId></Error>",
    "presigned_get_object_url": "http://72.28.97.51:31726/minio-py-test-3983798c-873d-45bf-ba41-050b3d8c00d3/2471cd25-f3b6-4845-89ba-31cb16ac3b1d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191112%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191112T100349Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=34cc3775ed9960162368d85adaa8da679020b0d21f26de8ae6c23fed45e4896b",
    "bucket_name": "minio-py-test-3983798c-873d-45bf-ba41-050b3d8c00d3",
    "response.status": 403
  }
}

Executed 11 out of 14 tests successfully.
```
Error: No such image: minio/minio:8510-54e661e

--- a/cmd/config/config.go
+++ b/cmd/config/config.go
@@ -405,6 +405,9 @@ func (c Config) SetKVS(s string, defaultKVS map[string]KVS) error {
 			kvs[prevK] = strings.Join([]string{kvs[prevK], sanitizeValue(kv[0])}, KvSpaceSeparator)
 			continue
 		}
+		if len(kv) == 1 {
+			return Error(fmt.Sprintf("key '%s', cannot have empty value", kv[0]))
+		}
 		prevK = kv[0]
 		kvs[kv[0]] = sanitizeValue(kv[1])
 	}

--- a/cmd/config/etcd/etcd.go
+++ b/cmd/config/etcd/etcd.go
@@ -95,6 +95,39 @@ func parseEndpoints(endpoints string) ([]string, bool, error) {
 	return etcdEndpoints, etcdSecure, nil
 }
 
+func lookupLegacyConfig(rootCAs *x509.CertPool) (Config, error) {
+	cfg := Config{}
+	endpoints := env.Get(EnvEtcdEndpoints, "")
+	if endpoints == "" {
+		return cfg, nil
+	}
+	etcdEndpoints, etcdSecure, err := parseEndpoints(endpoints)
+	if err != nil {
+		return cfg, err
+	}
+	cfg.Enabled = true
+	cfg.DialTimeout = defaultDialTimeout
+	cfg.DialKeepAliveTime = defaultDialKeepAlive
+	cfg.Endpoints = etcdEndpoints
+	cfg.CoreDNSPath = "/skydns"
+	if etcdSecure {
+		cfg.TLS = &tls.Config{
+			RootCAs: rootCAs,
+		}
+		// This is only to support client side certificate authentication
+		// https://coreos.com/etcd/docs/latest/op-guide/security.html
+		etcdClientCertFile := env.Get(EnvEtcdClientCert, "")
+		etcdClientCertKey := env.Get(EnvEtcdClientCertKey, "")
+		if etcdClientCertFile != "" && etcdClientCertKey != "" {
+			cfg.TLS.GetClientCertificate = func(unused *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				cert, err := tls.LoadX509KeyPair(etcdClientCertFile, etcdClientCertKey)
+				return &cert, err
+			}
+		}
+	}
+	return cfg, nil
+}
+
 // LookupConfig - Initialize new etcd config.
 func LookupConfig(kv config.KVS, rootCAs *x509.CertPool) (Config, error) {
 	cfg := Config{}
@@ -102,18 +135,35 @@ func LookupConfig(kv config.KVS, rootCAs *x509.CertPool) (Config, error) {
 		return cfg, err
 	}
 
-	stateBool, err := config.ParseBool(env.Get(EnvEtcdState, kv.Get(config.State)))
+	stateBool, err := config.ParseBool(env.Get(EnvEtcdState, config.StateOn))
 	if err != nil {
 		return cfg, err
 	}
 
-	endpoints := env.Get(EnvEtcdEndpoints, kv.Get(Endpoints))
-	if stateBool && len(endpoints) == 0 {
-		return cfg, config.Error("'endpoints' key cannot be empty if you wish to enable etcd")
+	if stateBool {
+		// By default state is 'on' to honor legacy config.
+		cfg, err = lookupLegacyConfig(rootCAs)
+		if err != nil {
+			return cfg, err
+		}
+		// If old legacy config is enabled honor it.
+		if cfg.Enabled {
+			return cfg, nil
+		}
 	}
 
-	if len(endpoints) == 0 && !stateBool {
+	stateBool, err = config.ParseBool(env.Get(EnvEtcdState, kv.Get(config.State)))
+	if err != nil {
+		return cfg, err
+	}
+
+	if !stateBool {
 		return cfg, nil
+	}
+
+	endpoints := env.Get(EnvEtcdEndpoints, kv.Get(Endpoints))
+	if endpoints == "" {
+		return cfg, config.Error("'endpoints' key cannot be empty to enable etcd")
 	}
 
 	etcdEndpoints, etcdSecure, err := parseEndpoints(endpoints)

--- a/cmd/gateway-main.go
+++ b/cmd/gateway-main.go
@@ -147,6 +147,25 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 	// Set when gateway is enabled
 	globalIsGateway = true
 
+	enableConfigOps := gatewayName == "nas"
+
+	// TODO: We need to move this code with globalConfigSys.Init()
+	// for now keep it here such that "s3" gateway layer initializes
+	// itself properly when KMS is set.
+
+	// Initialize server config.
+	srvCfg := newServerConfig()
+
+	// Override any values from ENVs.
+	if err := lookupConfigs(srvCfg); err != nil {
+		logger.FatalIf(err, "Unable to initialize server config")
+	}
+
+	// hold the mutex lock before a new config is assigned.
+	globalServerConfigMu.Lock()
+	globalServerConfig = srvCfg
+	globalServerConfigMu.Unlock()
+
 	router := mux.NewRouter().SkipClean(true)
 
 	if globalEtcdClient != nil {
@@ -156,8 +175,6 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 
 	// Initialize globalConsoleSys system
 	globalConsoleSys = NewConsoleLogger(context.Background(), globalEndpoints)
-
-	enableConfigOps := gatewayName == "nas"
 	enableIAMOps := globalEtcdClient != nil
 
 	// Enable IAM admin APIs if etcd is enabled, if not just enable basic
@@ -187,31 +204,13 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 		getCert = globalTLSCerts.GetCertificate
 	}
 
-	globalHTTPServer = xhttp.NewServer([]string{globalCLIContext.Addr}, criticalErrorHandler{registerHandlers(router, globalHandlers...)}, getCert)
+	globalHTTPServer = xhttp.NewServer([]string{globalCLIContext.Addr},
+		criticalErrorHandler{registerHandlers(router, globalHandlers...)}, getCert)
 	go func() {
 		globalHTTPServerErrorCh <- globalHTTPServer.Start()
 	}()
 
 	signal.Notify(globalOSSignalCh, os.Interrupt, syscall.SIGTERM)
-
-	if !enableConfigOps {
-		// TODO: We need to move this code with globalConfigSys.Init()
-		// for now keep it here such that "s3" gateway layer initializes
-		// itself properly when KMS is set.
-
-		// Initialize server config.
-		srvCfg := newServerConfig()
-
-		// Override any values from ENVs.
-		if err := lookupConfigs(srvCfg); err != nil {
-			logger.FatalIf(err, "Unable to initialize server config")
-		}
-
-		// hold the mutex lock before a new config is assigned.
-		globalServerConfigMu.Lock()
-		globalServerConfig = srvCfg
-		globalServerConfigMu.Unlock()
-	}
 
 	newObject, err := gw.NewGatewayLayer(globalActiveCred)
 	if err != nil {
@@ -252,7 +251,6 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 	// ****  WARNING ****
 	// Migrating to encrypted backend should happen before initialization of any
 	// sub-systems, make sure that we do not move the above codeblock elsewhere.
-
 	if enableConfigOps {
 		logger.FatalIf(globalConfigSys.Init(newObject), "Unable to initialize config system")
 

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -578,7 +578,12 @@ func (sys *IAMSys) ListUsers() (map[string]madmin.UserInfo, error) {
 	for k, v := range sys.iamUsersMap {
 		users[k] = madmin.UserInfo{
 			PolicyName: sys.iamUserPolicyMap[k].Policy,
-			Status:     madmin.AccountStatus(v.Status),
+			Status: func() madmin.AccountStatus {
+				if v.IsValid() {
+					return madmin.AccountEnabled
+				}
+				return madmin.AccountDisabled
+			}(),
 		}
 	}
 
@@ -609,8 +614,13 @@ func (sys *IAMSys) GetUserInfo(name string) (u madmin.UserInfo, err error) {
 
 	u = madmin.UserInfo{
 		PolicyName: sys.iamUserPolicyMap[name].Policy,
-		Status:     madmin.AccountStatus(creds.Status),
-		MemberOf:   sys.iamUserGroupMemberships[name].ToSlice(),
+		Status: func() madmin.AccountStatus {
+			if creds.IsValid() {
+				return madmin.AccountEnabled
+			}
+			return madmin.AccountDisabled
+		}(),
+		MemberOf: sys.iamUserGroupMemberships[name].ToSlice(),
 	}
 	return u, nil
 }

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -235,15 +235,6 @@ func initSafeModeInit(buckets []BucketInfo) (err error) {
 		return err
 	}
 
-	if globalEtcdClient != nil {
-		// ****  WARNING ****
-		// Migrating to encrypted backend on etcd should happen before initialization of
-		// IAM sub-systems, make sure that we do not move the above codeblock elsewhere.
-		if err = migrateIAMConfigsEtcdToEncrypted(globalEtcdClient); err != nil {
-			return fmt.Errorf("Unable to handle encrypted backend for iam and policies: %v", err)
-		}
-	}
-
 	return nil
 }
 
@@ -255,6 +246,15 @@ func initAllSubsystems(buckets []BucketInfo, newObject ObjectLayer) (err error) 
 
 	if err = globalNotificationSys.AddNotificationTargetsFromConfig(globalServerConfig); err != nil {
 		return fmt.Errorf("Unable to initialize notification target(s) from config: %w", err)
+	}
+
+	if globalEtcdClient != nil {
+		// ****  WARNING ****
+		// Migrating to encrypted backend on etcd should happen before initialization of
+		// IAM sub-systems, make sure that we do not move the above codeblock elsewhere.
+		if err = migrateIAMConfigsEtcdToEncrypted(globalEtcdClient); err != nil {
+			return fmt.Errorf("Unable to handle encrypted backend for iam and policies: %v", err)
+		}
 	}
 
 	if err = globalIAMSys.Init(newObject); err != nil {

--- a/pkg/madmin/parse-kv.go
+++ b/pkg/madmin/parse-kv.go
@@ -37,10 +37,23 @@ const (
 	commentKey = "comment"
 )
 
+// Count - returns total numbers of target
+func (t Targets) Count() int {
+	var count int
+	for _, targetKV := range t {
+		for range targetKV {
+			count++
+		}
+	}
+	return count
+}
+
 func (t Targets) String() string {
 	var s strings.Builder
+	count := t.Count()
 	for subSys, targetKV := range t {
 		for target, kv := range targetKV {
+			count--
 			c := kv[commentKey]
 			data, err := base64.RawStdEncoding.DecodeString(c)
 			if err == nil {
@@ -73,7 +86,7 @@ func (t Targets) String() string {
 				s.WriteString(KvDoubleQuote)
 				s.WriteString(KvSpaceSeparator)
 			}
-			if len(t) > 1 {
+			if (len(t) > 1 || len(targetKV) > 1) && count > 0 {
 				s.WriteString(KvNewline)
 				s.WriteString(KvNewline)
 			}
@@ -121,7 +134,7 @@ func convertTargets(s string, targets Targets) error {
 			kvs[prevK] = strings.Join([]string{kvs[prevK], sanitizeValue(kv[0])}, KvSpaceSeparator)
 			continue
 		}
-		if len(kv[1]) == 0 {
+		if len(kv) == 1 {
 			return fmt.Errorf("value for key '%s' cannot be empty", kv[0])
 		}
 		prevK = kv[0]

